### ISSUE TITLE
Crash in UCIFastReader

### ISSUE THREAD

**Issue by shane6lee** [Microsoft/CNTK#255]
I am seeing a crash in UCIFastReader and I think it is due to this line of code (UCIFastReader.cpp, line 928):

memcpy(&m_featuresBuffer.get()[j` \* m_featureCount], &m_featureData[jRand \* m_featureCount], sizeof(ElemType) \* m_featureCount);

There should be a check to ensure jRand \* m_featureCount + m_featureCount does not exceed the length of m_featureData.

**Comment by frankseide** [comment]
This was a known bug that should have been fixed some time ago. Would you mind checking two things:

* are you using the latest version? (Are you using a pre-built binary or self-compiled source?)
* did you declare SGD's epochSize parameter being larger than the actual training corpus? If so, please try adjusting epochSize = #samples in the corpus

**Comment by shane6lee** [comment]
I cloned the source from GitHub yesterday and compiled the binaries myself, so it is the latest version. In the SGD section the epochSize is set to 0.

**Comment by frankseide** [comment]
Would you mind trying something: If you know the number of samples in your training set, could you set epochSize to that number?

**Comment by shane6lee** [comment]
Actually while trouble shooting this error:
"EXCEPTION occurred: randomizeRange must be an even multiple of mbSize and an integral factor of epochSize",
I put in additional print statements and this was printed:
m_epochSize = 12277393; requestDataSize = 4611686018427387903; m_randomizeRange = 12277393; m_mbSize = 4096

So it looks like the epochSize is set to the number of input samples of 12277393.

**Comment by dongyu888** [comment]
it also requires the random range to be multiples of the minibatch size.
the workaround is to randomize your set offline and don't use readers randomization


Thanks,
Dong Yu (俞栋)Sent from my smart phone. Please forgive my typos.
<div>
</div><div>
</div><div>-------- Original message --------</div><div>From: shane6lee <notifications@github.com> </div><div>Date: 3/20/2016  13:17  (GMT+08:00) </div><div>To: Microsoft/CNTK <CNTK@noreply.github.com> </div><div>Subject: Re: [CNTK] Crash in UCIFastReader (#255) </div><div>
</div>
Actually while trouble shooting this error:
"EXCEPTION occurred: randomizeRange must be an even multiple of mbSize and an integral factor of epochSize",
I put in additional print statements and this was printed:
m_epochSize = 12277393; requestDataSize = 4611686018427387903; m_randomizeRange = 12277393; m_mbSize = 4096

So it looks like the epochSize is set to the number of input samples of 12277393.

---

**Comment by raaaar** [comment]
If you're feeling adventurous, you can try using CNTKTextFormat reader instead of the UCIFast. 
Towards this end, you'll need to 

- convert your data using a python [script](https://github.com/Microsoft/CNTK/blob/alrezni/TextReaderExperimental/Source/Readers/CNTKTextFormatReader/uci_to_cntk_text_format_converter.py): 

`python uci_to_cntk_text_format_converter.py -in [input file] -fs [features.start] -fd [features.dim] -ls [labels.start] -ld [labels.dim] --num_labels [labels.labelDim] -out [output file]` 
where input values are identical to those in your "reader" config section.

- replace "reader" section in you config with: 

```
reader = [
        readerType = "CNTKTextFormatReader"
        file = "[output file]" # output of the converter
		input = [
			features = [
				dim = [features.dim] # same as before
				format = "dense"
			]
			labels = [
				dim = [labels.labelDim] # the number of possible label values (not labels.dim!)
				format = "dense"
			]
		]
    ]
```

**Comment by shane6lee** [comment]
Using the "Simple2d" example, could you tell me if this is the right way to use the conversion script:

`python uci_to_cntk.py -in SimpleDataTrain.txt -fs 0 -fd 2 -ls 2 -ld 1 --num_labels 2 -out cntk.txt`

**Comment by raaaar** [comment]
That is correct. The reader section in the config should look like this:
```
reader = [
        readerType = "CNTKTextFormatReader"
        file = "cntk.txt" # alternatively, "full_path/cntk.txt" or "$DataDir$/cntk.txt"
		
        input = [
            features = [
                dim = 2
                format = "dense"
            ]
            labels = [
                dim = 2
                format = "dense"
            ]
        ]
    ]
```

**Comment by pkranen** [comment]
The readers have bee reworked entirely. Please try with the latest bits. Thanks.
